src/replacer.rs

```diff
@@ -0,0 +1,361 @@
+    // Algorithm: call finder.find(). If it matches, bump input_index and output_text
+    // using the length of the input text and the length of the output text respectively
+    // Truncate the input string at each step to keep finding successive matches:
+    //
+    // step | buf                      | input_fragment | output_fragent
+    // -----|--------------------------|----------------|----------
+    //    0 | "my tea is the best tea" |     (3, "tea") | (3, "coffe")
+    //    1 | "  is the best tea"      |    (19, "tea") | (20, "coffe")
+    //    2 | " !"                     |      n/a       | n/a
+    //
+    let mut fragments = Fragments::new();
+    let mut input_index = 0;
+    let mut output_index = 0;
+    let mut buff = input;
+    while let Some(res) = finder.replace(buff) {
+        let (pos, input_text, output_text) = res;
+        input_index += pos;
+        output_index += pos;
+        fragments.add((input_index, &input_text), (output_index, &output_text));
+        let new_start = input_index + input_text.len();
+        buff = &input[new_start..];
+        input_index += input_text.len();
+        output_index += output_text.len();
+    }
+
+    fragments
+}
+
+fn get_output(input: &str, fragments: &Fragments) -> String {
+    let mut current_index = 0;
+    let mut output = String::new();
+    for (input_fragment, output_fragment) in fragments.into_iter() {
+        let Fragment {
+            text: input_text,
+            index: input_index,
+        } = input_fragment;
+
+        let Fragment {
+            text: output_text, ..
+        } = output_fragment;
+
+        output.push_str(&input[current_index..*input_index]);
+        output.push_str(&output_text);
+        current_index = input_index + input_text.len();
+    }
+    output.push_str(&input[current_index..]);
+    output
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+    use regex::Regex;
+
+    #[test]
+    fn test_substring_1() {
+        let input = "Mon thé c'est le meilleur des thés !";
+        let pattern = "thé";
+        let replacement = "café";
+        let query = Query::substring(pattern, replacement);
+        let replacement = replace(input, &query).unwrap();
+        assert_eq!(
+            replacement.output(),
+            "Mon café c'est le meilleur des cafés !"
+        );
+    }
+
+    #[test]
+    fn test_substring_2() {
+        let input = "old old old";
+        let pattern = "old";
+        let replacement = "new";
+        let query = Query::substring(pattern, replacement);
+        let replacement = replace(input, &query).unwrap();
+        assert_eq!(replacement.output(), "new new new");
+    }
+
+    #[test]
+    fn test_display_patch() {
+        let input = "Top: old is nice";
+        let pattern = "old";
+        let replacement = "new";
+        let query = Query::substring(pattern, replacement);
+        let replacement = replace(input, &query).unwrap();
+        replacement.print_self("foo.txt:3 ");
```

Comment:
where is the assert ?  :D

Reply:
not really needed - the test is there just so that it's easier to tweak the output